File: 143.重排链表.cpp
/*
 * @lc app=leetcode.cn id=143 lang=cpp
 *
 * [143] 重排链表
 */

// @lc code=start

#include "iostream"
#include "vector"
#include "map"

using namespace std;
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    void reorderList(ListNode* head) {
        // 找到中点
        // 反转后面的链表
        // 合并两个链表
        if(!head || !head->next) {
            return;
        }
        ListNode *p = head;
        ListNode *q = head;

        while(q && q->next && q->next->next) {
            q = q->next->next;
            p = p->next;
        }
        q = p;
        ListNode *r = reverseList(p->next);
        merge(head, r, q);
    }

    void merge(ListNode *l, ListNode *r, ListNode *mid) {
        if (!l || !r) return;
        while(l != mid && r) {
            ListNode *t = l;
            ListNode *lt = l->next;
            ListNode *rt = r->next;
            l->next = r;
            r->next = lt;
            l = lt;
            r = rt;
        }
        l->next = r;
    }

     ListNode *reverseList(ListNode *head) {
        if (!head || !head->next)
            return head;
        ListNode *cur = head;
        ListNode *next = head->next;
        cur -> next = NULL;
        ListNode *p = cur->next, *q = next->next;
        while (next) {
            cur->next = p;
            q = next->next;
            next->next = cur;
            p = cur;
            cur = next;
            next = q;
        }
        return cur;
    }


};
// @lc code=end

int main() {
    Solution s;
    ListNode l5(5);
    ListNode l4(4, &l5);
    ListNode l3(3, &l4);
    ListNode l2(2, &l3);
    ListNode l1(1, &l2);
    s.reorderList(&l1);
}

File: 143.重排链表.go
/*
 * @lc app=leetcode.cn id=143 lang=golang
 *
 * [143] 重排链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

 func reverse(head *ListNode) *ListNode {
	cur, next := head, head.Next
	var pre *ListNode = nil
	for cur != nil {
		next = cur.Next
		cur.Next = pre
		pre = cur
		cur = next
	}
	return pre
}

func merge(left, right *ListNode) {
	var pl, pr *ListNode
	for left != nil && right != nil {
		ln := left.Next
		rn := right.Next
		left.Next = right
		right.Next = ln
		pl = left
		pr = right
		right = rn
		left = ln
	}
	if right != nil {
		pr.Next = right
	}
	if left != nil {
		pl.Next = left
	}
}


// 快慢指针，先定位到中点， 再反转后面部分，再合并两个列表
func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}
	slow, quick := head, head
	pre := quick
	for quick != nil && quick.Next != nil {
		pre = slow
		slow = slow.Next
		quick = quick.Next.Next
	}
	pre.Next = nil
	quick = reverse(slow)
	merge(head, quick)
}
// @lc code=end

